plat/kvm/x86/traps.c

```diff
@@ -151,19 +151,77 @@ static void tss_init(__lcpuidx idx)
 		: "r"((__u16) (GDT_DESC_OFFSET(GDT_DESC_TSS_LO))));
 }
 
+static __align(8)
+UKPLAT_PER_LCPU_ARRAY_DEFINE(struct seg_gate_desc64, cpu_idt, IDT_NUM_ENTRIES);
+
+static
+UKPLAT_PER_LCPU_DEFINE(__u8, idt_ist_disable_nesting);
+
+#define IDT_IST_SAVE_LEN 32
+/* Space for the IST values of all exception vectors */
+static __u8 idt_ist_saved[IDT_IST_SAVE_LEN];
+
+void ukarch_push_nested_exceptions(void)
+{
+	struct seg_gate_desc64 *desc;
+	struct seg_gate_desc64 *idt;
+	__u8 *disable_nesting;
+	unsigned int i;
+
+	disable_nesting = &ukplat_per_lcpu_current(idt_ist_disable_nesting);
+	UK_ASSERT(*disable_nesting < __U8_MAX);
+
+	if (*disable_nesting++)
+		return;
+
+	idt = ukplat_per_lcpu_current(cpu_idt);
```

Comment:
This might prove to be unsafe as you cannot know for sure if you can safely access the lcpu idx field, given that it is an arbitrary section that does not guarantee the state of gs_base. Perhaps we could use `lcpu_get_current_in_except` to get the index and then reference the variable through `ukplat_per_lcpu`.